### ISSUE TITLE
Improve Nav Editor placeholder state

### ISSUE THREAD

**Issue by noisysocks** [WordPress/gutenberg#23953]
Related: https://github.com/WordPress/gutenberg/issues/23207

Right now when you use the new Navigation screen to create a new menu, you are shown the Navigation block's placeholder which prompts you to create from an existing WordPress menu. This list contains the menu that you just created.

Could we simplify this flow for the most common case which is creating a new empty menu?

Initial placeholder state:
<img width="467" alt="Screenshot 2021-01-12 at 1 53 31 pm" src="https://user-images.githubusercontent.com/677833/104275352-a2287c80-54dd-11eb-96f8-4dac06caf9a0.png">

On selection of the nav block:
<img width="461" alt="Screenshot 2021-01-12 at 1 54 40 pm" src="https://user-images.githubusercontent.com/677833/104275429-c5532c00-54dd-11eb-88c7-8868163d4b8a.png">

As recommended (https://github.com/WordPress/gutenberg/issues/28124#issuecomment-765702302) it could look like this:

https://user-images.githubusercontent.com/677833/109124595-11a0b580-7786-11eb-9bb8-6e6a450bc5bd.mp4

**Comment by talldan** [comment]
The screen and block have changed quite a bit, but this issue is still relevant. I've updated the description with some new screenshots.

The placeholder has been redesigned with full-site editing in mind. It now shows an 'empty state', which doesn't work so well in the navigation screen.

**Comment by jasmussen** [comment]
There's only ever a signle navigation block per "Menu", correct? If that's the case, can we make it appear selected always? I.e. force either `is-selected`, `has-child-selected`, or both?

That would skip the skeleton preview state.

**Comment by talldan** [comment]
Yep, we can definitely make a few customisations in the context of the screen to make the skeleton bit never show. The other issue that's harder to solve is the current menu appearing in the 'Existing' dropdown. More of a technical issue though. Let's move this one over to the Needs Dev column.

**Comment by talldan** [comment]
There's a design for this now - https://github.com/WordPress/gutenberg/issues/28124#issuecomment-765702302

It's a completely reworked placeholder, so we'll have to look into how we can replace the placeholder in the nav block only in the context of the nav screen.

**Comment by jasmussen** [comment]
Question: Is the design in https://github.com/WordPress/gutenberg/issues/28124#issuecomment-765702302 actually a _placeholder state_? I thought it was UI that existed before any block was inserted, and not part of the Navigation Block itself. As always I will defer to you all, but it seems best if we can avoid building in navigation screen special cases into the navigation block itself.

**Comment by talldan** [comment]
Very true, that's worth exploring too.

**Comment by talldan** [comment]
So I think @jasmussen's idea above makes sense.

Essentially the nav screen could provide its own UI and we'd be deferring creation of the nav block until after 'Add all pages' or 'Start blank' is clicked.

I think there will have to be a little bit of hacking on the block to make the 'Start blank' option work (otherwise choosing it will show the block with its normal placeholder).

Add a hasPlaceholder prop for the component:
https://github.com/WordPress/gutenberg/blob/57342da8a3596b0b54c04e31e505833fa4bf7227/packages/block-library/src/navigation/edit.js#L44

And use that to default isPlaceholderShown to false, and also make this placeholder prop undefined:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation/edit.js#L91

**Comment by draganescu** [comment]
One other thing is that Add all pages will have to behave like it did before the pages block, since classic menus don't need this to be self updating, plus the editor won't support the block without the theme optin.

**Comment by getdave** [comment]
@critterverse Are you happy with the approach shown in the video in the description? If so I can take this on.

**Comment by javierarce** [comment]
I'll answer for @critterverse since she is working on the widget editor, and I think this mention was meant for me (but @critterverse, please jump in if you have a different idea).

I like the approach shown in the video and @jasmussen's suggestion of skipping the skeleton preview state. 
Regarding the "Existing menu" dropdown… would we get rid of it, or do you want me to add it to the design?

**Comment by draganescu** [comment]
> do you want me to add it to the design?

That would add a bit of more clarity which would be helpful :)

**Comment by javierarce** [comment]
Here is the modal with the "Existing menu" dropdown. I've also updated the order of the options so "start blank" appears in the first position (which seems more logical to me) and remove the down arrow.

<img width="1152" alt="Placeholder" src="https://user-images.githubusercontent.com/4933/116065136-fa4b4f00-a686-11eb-8035-a7ffb858239c.png"> 

I also think it would be nice to use the same header style in both modals:

<img width="1152" alt="Placeholder" src="https://user-images.githubusercontent.com/4933/115901483-8b40e100-a461-11eb-9f2f-58c67f4afdf4.png">

cc: @shaunandrews

**Comment by getdave** [comment]
@javierarce Let me know when this is good to go and I'll start work on a PR.

**Comment by javierarce** [comment]
Thanks, @getdave! Let me get @shaunandrews opinion first.

Here are the designs for the mobile version. I know that we currently don't have padding around the main canvas, but I'd like to propose adding it (I'll open a new issue for that).

<img width="375" alt="Mobile" src="https://user-images.githubusercontent.com/4933/116065289-28c92a00-a687-11eb-8c24-eb2daa16d97b.png">

<img width="375" alt="Mobile-1" src="https://user-images.githubusercontent.com/4933/116065284-28309380-a687-11eb-886e-e053b2ff5fd7.png">

**Comment by shaunandrews** [comment]
> I also think it would be nice to use the same header style

I'm not sure were we came up with the line, but I think it was likely from a mockup that used an accordion. I'm not sure the line is necessary.

--

Why is "Add existing menu" a primary button? Not all people will have existing menus to choose from...

**Comment by javierarce** [comment]
You are right about the line, I got it from a mockup. I've removed it.

> Why is "Add existing menu" a primary button? Not all people will have existing menus to choose from...

That's true. I've updated the design so the button doesn't look so prominent and behaves like the _Switch menu_ button (for that reason I think we don't need the arrow).

In case the user doesn't have other menus, we won't show that option (this is by the way related to [this bug]( https://github.com/WordPress/gutenberg/issues/30745): the menu itself is always offered as an option in the existing menu dropdown)

<img width="1152" alt="Placeholder" src="https://user-images.githubusercontent.com/4933/116214858-6b523b80-a747-11eb-9ea4-f9f77fc833b1.png">

<img width="1152" alt="Placeholder-2" src="https://user-images.githubusercontent.com/4933/116214856-6ab9a500-a747-11eb-8e83-4c70104e8afc.png">


<img width="1152" alt="Placeholder-1" src="https://user-images.githubusercontent.com/4933/116214853-6a210e80-a747-11eb-9963-c093057ed02c.png">

<img width="375" alt="Mobile" src="https://user-images.githubusercontent.com/4933/116214849-6a210e80-a747-11eb-9b69-d9dda66fab24.png"> <img width="375" alt="Mobile-1" src="https://user-images.githubusercontent.com/4933/116214846-68efe180-a747-11eb-9666-2abe64abc52c.png">

**Comment by getdave** [comment]
@gwwar If we were to implement the design shown in this Issue I believe we are going to start duplicating a lot of code between

* Nav block
* Nav editor

Why? This PR is effectively implementing the Nav _block_ placeholder state in the Nav editor. That means we'll need to duplicate all the code for:

* fetching existing pages and parsing into blocks.
* fetching menus and parsing into blocks.

Are we reaching a point where we need to place such code within a new `@wordpress/navigation` package and then share it between both the block and the editor? That will reduce redundancy and make maintenance easier.

I'm loath to start a major refactor but I'm interested in your opinion on alternatives before we start adding yet more code duplication.

I was thinking perhaps we could add modifiers to the block itself to allow for a simpler placeholder state but I don't believe we should start coupling the block and editor too tightly.

**Comment by jasmussen** [comment]
Regarding the nav block placeholder state, I'm afraid it remains a bit in flux. The skeleton indicators remain a source of confusion, even if small, and I've noticed an issue where the nav block when inserted in constrained contexts gets scaled improperly, so the nav block placeholder needs a bit of a refactor itself. 

So it might not need to be that much of a duplication. If we're lucky, it can be a good exercise in mostly using stock components — paragraph, tertiary and primary buttons, input fields, and as little other CSS as possible.

**Comment by getdave** [comment]
Just to be clear (apologies that I wasn't!), the duplication I'm primarily concerned about is the code that queries and formats the menu/pages data into blocks. The UI is less of an issue in this case although I can imagine quite a lot of the component-level logic from [placeholder.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/placeholder.js) being repeated in the Nav Editor if we go down this placeholder route.

Here are some examples of the code I'd like to reuse in the Nav Editor that currently only exists in the block:

* https://github.com/WordPress/gutenberg/blob/e9017989d173fb41a1f3e483c70e721c279e6d89/packages/block-library/src/navigation/create-data-tree.js#L16-L37
* https://github.com/WordPress/gutenberg/blob/e9017989d173fb41a1f3e483c70e721c279e6d89/packages/block-library/src/navigation/placeholder.js#L40-L83

**Comment by gwwar** [comment]
> Are we reaching a point where we need to place such code within a new @wordpress/navigation package and then share it between both the block and the editor? That will reduce redundancy and make maintenance easier.

@getdave That makes sense to me. I'm not sure on the package name, but likely there's going to be a need for similar code in all editors: post, site, widget, navigation. We saw this with the shared search function.

**Comment by getdave** [comment]
Looks like I need to raise an Issue/PR combo for this then.

**Comment by talldan** [comment]
'Add existing menu' is an interesting option on this screen, because it is the menu screen where these menus are created/edited. 

I think it'd feel more like copying or duplicating an existing menu in this context.

> @getdave That makes sense to me. I'm not sure on the package name, but likely there's going to be a need for similar code in all editors: post, site, widget, navigation. We saw this with the shared search function.

A `widgets` package was recently added for sharing code between the customizer and standalone widget editors. So a `menus` package could be similar.

I also wonder if we're reaching the point where `block-library` has too much in it. It has generic use everywhere blocks (paragraph/quote etc.). As well as WordPress specific blocks (latest posts). That makes me wonder if there should be `wordpress/blocks` and `wordpress/wp-blocks` packages, where the latter is not too concerned about having WordPress-y dependencies.

**Comment by jasmussen** [comment]
Interesting the you should mention "Add existing menu", as https://github.com/WordPress/gutenberg/pull/31371 changes the phrasing to that in the block setup state.

For the block, the motivation is to make clear that it's just a starting point, a copy if you will, and that there's no connection between screen and block after that fact. Honestly, "Copy menu" might be even better phrasing to help suggest that.

**Comment by getdave** [comment]
Here's the proposal for a `@wordpress/menus` package.

https://github.com/WordPress/gutenberg/issues/31580

**Comment by kellychoffman** [comment]
Locking this issue as design is ready for dev. "Locking" is for dev clarity and is not meant to dissuade further design contribution or discussion - open new issues or unlock if needed.

[Comment containing final design](https://github.com/WordPress/gutenberg/issues/23953#issuecomment-827446427)
[Figma file](https://www.figma.com/file/xvnD8SOGhQYvnHJLGXYWGJ/?node-id=530%3A194) 
Designer to add to PR for design reviews: @javierarce

**Comment by talldan** [comment]
@kellychoffman I'm not sure locking the issue is a good practice as a lot of contributors like to leave a comment volunteering to pick up an issue. That's a good idea as it prevents duplicate efforts.

I've unlocked it for now, could be worth a wider discussion on the practice.
